api/src/main/java/bio/terra/tanagra/underlay/displayhint/EnumVals.java

```diff
@@ -0,0 +1,183 @@
+  private static final Logger LOGGER = LoggerFactory.getLogger(EnumVals.class);
+  private static final String ENUM_VALUE_COLUMN_ALIAS = "enumVal";
+  private static final int MAX_ENUM_VALS_FOR_DISPLAY_HINT = 100;
+
+  private final List<ValueDisplay> valueDisplays;
+
+  public EnumVals(List<ValueDisplay> valueDisplays) {
+    this.valueDisplays = valueDisplays;
+  }
+
+  public static EnumVals fromSerialized(UFEnumVals serialized) {
+    if (serialized.getValueDisplays() == null) {
+      throw new IllegalArgumentException("Enum values map is undefined");
+    }
+    List<ValueDisplay> valueDisplays =
+        serialized.getValueDisplays().stream()
+            .map(vd -> ValueDisplay.fromSerialized(vd))
+            .collect(Collectors.toList());
+    return new EnumVals(valueDisplays);
+  }
+
+  @Override
+  public Type getType() {
+    return Type.ENUM;
+  }
+
+  public List<ValueDisplay> getValueDisplays() {
+    return Collections.unmodifiableList(valueDisplays);
+  }
+
+  public static EnumVals computeForField(Literal.DataType dataType, FieldPointer value) {
+    List<TableVariable> tables = new ArrayList<>();
+    TableVariable primaryTable = TableVariable.forPrimary(value.getTablePointer());
+    tables.add(primaryTable);
+
+    final String enumValAlias = "enumVal";
+
+    FieldVariable valueFieldVar = value.buildVariable(primaryTable, tables, enumValAlias);
+    Query query =
+        new Query.Builder()
+            .select(List.of(valueFieldVar))
+            .tables(tables)
+            .orderBy(List.of(valueFieldVar))
+            .groupBy(List.of(valueFieldVar))
+            .limit(MAX_ENUM_VALS_FOR_DISPLAY_HINT + 1)
+            .build();
+
+    List<ColumnSchema> columnSchemas =
+        List.of(
+            new ColumnSchema(enumValAlias, CellValue.SQLDataType.fromUnderlayDataType(dataType)));
+
+    DataPointer dataPointer = value.getTablePointer().getDataPointer();
+    QueryRequest queryRequest =
+        new QueryRequest(query.renderSQL(), new ColumnHeaderSchema(columnSchemas));
+    QueryResult queryResult = dataPointer.getQueryExecutor().execute(queryRequest);
+
+    List<String> enumStringVals = new ArrayList<>();
+    Iterator<RowResult> rowResultIter = queryResult.getRowResults().iterator();
+    while (rowResultIter.hasNext()) {
+      enumStringVals.add(rowResultIter.next().get(enumValAlias).getString().orElse(null));
+      if (enumStringVals.size() > MAX_ENUM_VALS_FOR_DISPLAY_HINT) {
+        // if there are more than the max number of values, then skip the display hint
+        LOGGER.info(
+            "Skipping enum values display hint because there are >{} possible values: {}",
+            MAX_ENUM_VALS_FOR_DISPLAY_HINT,
+            valueFieldVar.getAlias());
+        return null;
+      }
+    }
+    return new EnumVals(
+        enumStringVals.stream().map(esv -> new ValueDisplay(esv)).collect(Collectors.toList()));
+  }
+
+  public static EnumVals computeForField(
```

Comment:
Can the two versions of this function be refactored to share code?

Reply:
Yes, done.